### PR TITLE
Remove some more unecessary cloning in sql_expr_to_logical_expr

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
Re https://github.com/apache/arrow-datafusion/issues/1469
Follow on to https://github.com/apache/arrow-datafusion/pull/1945# from @doki23  ❤️ 

# Changes
Remove some more unnecessary `clone()` while building up `Exprs` from `SQLExpr`. 

I wanted a small vanity project to make some actual code changes rather than filing ticket or reviewing more PRs lol -- see https://github.com/apache/arrow-datafusion/pull/1945#pullrequestreview-903442421

I'll highlight the places below that this isn't just mechanical

### CODE REVIEW

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -1495,121 +1493,134 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            } => self.parse_sql_binary_op(*left, op, *right, schema),
 
             #[cfg(feature = "unicode_expressions")]
             SQLExpr::Substring {
                 expr,
                 substring_from,
                 substring_for,
             } => {
-                let arg = self.sql_expr_to_logical_expr(expr, schema)?;
+                if substring_from.is_none() && substring_for.is_none() {
```
[the remark this answers is not in the document]
Comment (Dandandan) [reply datafusion/src/sql/planner.rs:1584]: If we want to make it cleaner (avoiding `unreachable` which I think is not very nice).

1. match on substring to create args based on from_logic, for_logic 
2. error on None/None
3. Push creation of `arg` to after match to produce `args`

**Review by Dandandan — APPROVED** [review]
Looks good, 1 comment about avoiding the unreachable match / panic.

File: datafusion/src/sql/planner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/src/sql/planner.rs:1584]: Thanks for the suggestion -- I actually tried this and found a challenge that `arg` is actually used in the match statements (to form a `Vec`). I could make a `mut vec` and push arg on the front afterwards, or do some iterator magic, but I felt in the end this was perhaps the easiest to understand formulation 🤔 

```rust
                        vec![arg, from_logic, for_logic]
```

**Comment by doki23** [comment]
👍🏻

File: datafusion/src/sql/planner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (houqp) [reply datafusion/src/sql/planner.rs:1584]: how about returning a result from the match statement? i.e. something like below:

```rust
            let args = match (substring_from, substring_for) {
                    (Some(from_expr), Some(for_expr)) => {
                        let from_logic =
                            self.sql_expr_to_logical_expr(*from_expr, schema)?;
                        let for_logic =
                            self.sql_expr_to_logical_expr(*for_expr, schema)?;
                        Ok(vec![arg, from_logic, for_logic])
                    }    
                    (None, None) => {
                            let orig_sql = SQLExpr::Substring {
                                expr,
                                substring_from: None,
                                substring_for: None,
                            };

                            Err(DataFusionError::Plan(format!(
                                "Substring without for/from is not valid {:?}",
                                orig_sql
                            )))
                    }
            }?;
```

**Review by houqp — APPROVED** [review]

**Review by yjshen — APPROVED** [review]
LGTM. Thanks @alamb

**Comment by alamb** [comment]
Thanks @yjshen !

**Pull request merged**